timers/timers.go

```diff
@@ -0,0 +1,224 @@
+	pad [cacheline - unsafe.Sizeof(sync.Mutex{}) - unsafe.Sizeof(&Timer{})]byte
+}
+
+// A Wheel schedules and executes deferred operations.
+type Wheel struct {
+	clock       clock.Clock
+	period      time.Duration
+	ticker      *clock.Ticker
+	bucketsMask uint64
+	buckets     []bucket
+	fired       chan *Timer
+	stopOnce    sync.Once
+	stop        chan struct{}
+	stopped     chan struct{}
+}
+
+// NewWheel creates and starts a new Wheel.
+func NewWheel(period, maxTimeout time.Duration) *Wheel {
+	return newWheel(period, maxTimeout, clock.New())
+}
+
+func newWheel(period, maxTimeout time.Duration, clock clock.Clock) *Wheel {
+	numBuckets, power := nextPowerOfTwo(int64(maxTimeout / period))
+	w := &Wheel{
+		clock:       clock,
+		period:      period,
+		ticker:      clock.Ticker(period),
+		bucketsMask: power,
+		buckets:     make([]bucket, numBuckets),
+		fired:       make(chan *Timer),
+		stop:        make(chan struct{}),
+		stopped:     make(chan struct{}),
+	}
+	go w.tick()
+	go w.work()
+	return w
+}
+
+// Stop shuts down the wheel, blocking until all the background goroutines
+// complete.
+func (w *Wheel) Stop() {
+	// TChannel Channels are safe to close more than once, so wheels should be
+	// too.
+	w.stopOnce.Do(func() {
+		w.ticker.Stop()
+		close(w.stop)
+		<-w.stopped
+		todo := w.gatherTimers(0, uint64(len(w.buckets)))
+		w.fire(todo)
+	})
+}
+
+// AfterFunc creates a Timer in the wheel.
+func (w *Wheel) AfterFunc(d time.Duration, f func()) *Timer {
+	deadline := w.asTick(w.clock.Now().Add(d))
+	t := &Timer{
+		f:        f,
+		deadline: deadline,
+	}
+
+	bucket := &w.buckets[deadline&w.bucketsMask]
+	bucket.Lock()
+	bucket.timers = bucket.timers.push(t)
+	bucket.Unlock()
+	return t
+}
+
+func (w *Wheel) tick() {
+	watermark := w.asTick(w.clock.Now())
+	for {
+		select {
+		case now := <-w.ticker.C:
+			nowTick := w.asTick(now)
+			todo := w.gatherTimers(watermark, nowTick)
+			if todo != nil {
+				w.fired <- todo
+			}
+			watermark = nowTick
+		case <-w.stop:
+			close(w.fired)
+			return
+		}
+	}
+}
+
+func (w *Wheel) work() {
+	for batch := range w.fired {
+		w.fire(batch)
+	}
+	close(w.stopped)
+}
+
+func (w *Wheel) fire(batch *Timer) {
+	for t := batch; t != nil; t = t.next {
+		t.f()
+	}
+}
+
+func (w *Wheel) gatherTimers(startTick, endTick uint64) *Timer {
+	var todo *Timer
+	for tick := startTick; tick < endTick; tick++ {
+		todo = w.gatherTimersForTick(tick, todo)
+	}
+	return todo
+}
+
+func (w *Wheel) gatherTimersForTick(tick uint64, todo *Timer) *Timer {
+	bucket := &w.buckets[tick&w.bucketsMask]
+
+	bucket.Lock()
+	batch := bucket.timers
+	bucket.timers = nil
+	bucket.Unlock()
+
+	var unexpired *Timer
+	for batch != nil {
+		next := batch.next
+		if batch.f != nil && batch.deadline > tick {
+			unexpired = unexpired.push(batch)
+		} else if batch.f != nil && batch.expire() {
+			todo = todo.push(batch)
+		}
+		batch = next
+	}
+	if unexpired != nil {
+		// We should only hit this case if we're a full wheel rotation
```

Comment:
We can certainly get metrics out; by design, though, locking the bucket won't help much, since most timers are going into some other bucket with a different lock. We can introduce a global synchronization point, but I don't think that's worth the cost.